src/plugin_output_processing/settings.py

```diff
@@ -44,18 +48,49 @@ def check_model(self):
         and can be called, it will be used as the default provider. If not, OpenAI will
         be used if the API key is set. Otherwise, the service will not start.
         """
-        for provider in [Ollama(), OpenAI()]:
-            if self.provider and provider.name != self.provider:
-                continue
-            if provider.available:
-                self.provider = provider.name
-                self.model = provider.get_model(self.model)
-                self.url = provider.url
-                logger.debug(
-                    f"Provider set to {self.provider} with model {self.model}."
+        if self.provider == OLLAMA_NAME or self.provider is None:
+            ollama = Ollama(self.model)
+            if ollama.available:
+                return self.set_provider(ollama)
+            if self.provider is not None:
+                logger.warning(
+                    f"Ollama provider not available, falling back to {OPENAI_NAME}."
+                )
+                self.provider = None
+
+        if self.provider == OPENAI_NAME or self.provider is None:
+            openai = OpenAI(self.model)
+            if openai.available:
+                return self.set_provider(openai)
+            # If OpenAI is set in the configuration but not available, ollama would not
+            # have been tried yet. So by recursing, we can try to use ollama as a fallback.
+            if self.provider is not None:
+                logger.warning(
+                    f"OpenAI provider not available, falling back to {OLLAMA_NAME}."
```

Comment:
Agree for more extensible, but in this case we have the same problem I had with the `for` loop, meaning that if you have a configuration with openai defined but you removed the token and you have an ollama instance, the loop will cause a provider error while my approach will correctly switch to ollama.

So it all comes down to: is it more likely to add a new provider or to change the configuration.

My thought is that it is more likely someone wants to try it with openai, see that it works and launch ollama while removing the token from openai.

Reply:
I understood the problem with the for loop but I think it doesn't occur with the while loop. 

Let's have a concret exemple.

If the provider is set to `openai` and the `OPENAI_API_KEY` is not set so, at the first iteration of the loop, `provider` is set to `OpenAI` and is removed from the dict `providers`. This provider not beeing available, at the second iteration `provider` is set to `None` because `openai` is no more a key of the dict `providers`. So provider is therefore set to the first item of this dict, in our case, `Ollama`.

Reply:
Let's have a try with your solution:

```
# PASS - Default to openai, I have a token, everything works as expected
docker compose up
...
api  | 2024-07-17 15:56:28.728 | INFO     | plugin_output_processing.settings:check_model:57 - Provider set to openai with model gpt-4o.
...

# FAIL - I just switch to ollama, it does not use ollama
docker compose down
docker compose --profile ollama up
...
api     | 2024-07-17 15:57:37.062 | WARNING  | plugin_output_processing.provider:get_model:49 - gpt-4o not found in model list with backend ollama, using default instead.
api     | 2024-07-17 15:58:11.885 | INFO     | plugin_output_processing.settings:check_model:57 - Provider set to openai with model gpt-4o.
...

# PASS - I unset the API token from openai and try without ollama backend, it works fine
unset OPENAI_API_KEY
docker compose up
...
api  | 2024-07-17 16:00:05.513 | WARNING  | plugin_output_processing.settings:check_model:64 - openai provider not available, falling back to the next one.
api  | 2024-07-17 16:00:05.513 | WARNING  | plugin_output_processing.settings:check_model:64 - ollama provider not available, falling back to the next one.
api  | 2024-07-17 16:00:05.513 | ERROR    | plugin_output_processing.settings:check_model:68 - None of the providers are available.

# PASS - the API key is still unset but I try a ollama backend
docker compose --profile ollama
...
api     | 2024-07-17 16:01:48.491 | WARNING  | plugin_output_processing.settings:check_model:64 - openai provider not available, falling back to the next one.
api     | 2024-07-17 16:01:48.491 | INFO     | plugin_output_processing.settings:check_model:57 - Provider set to ollama with model qwen2:0.5b.
...

# -----------

# Now let's forget docker a little bit and work directly by calling the API with uvicorn

# FAIL - Launch the application with default parameters. 2 problems:
#   1. The default backend is openai, not ollama
#   2. The backend exists but the model is incorrect, I would have like here that we replaced `qwen2:0.5b` by `gpt4-o`
docker start ollama
uvicorn --reload src.plugin_output_processing.api:app
2024-07-17 12:13:32.904 | WARNING  | src.plugin_output_processing.provider:get_model:49 - qwen2:0.5b not found in model list with backend openai, using default instead.
2024-07-17 12:13:32.905 | INFO     | src.plugin_output_processing.settings:check_model:57 - Provider set to ollama with model qwen2:0.5b
```

Reply:
What we need to make sure is:

1. If a configuration file is provided with a provider and a model, we should tests those in priority (idealy without having to ping other providers)
2. If the configuration file is incorrect (model not found in the list), we should try to find a model that works with this provider 
3. If the provider is not available, only then we can try another one
4. If there are no configurations, then we should try ollama then openai

Reply:
Accepted your change with few modifications:

- changed the order of the dictionary to put ollama last so it is tested first on a new configuration
- only instantiate the object when the provider is set and not within the dictionary

Reply:
I don't really understand the condition of the while loop. Why `[OLLAMA_NAME, OPENAI_NAME]` and not `providers` ?